### ISSUE TITLE
Required checks on GitHub Actions

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#26890]
Quite a few of the checks on GitHub Actions are no longer set as "required" for merging a PR, including the actual tests and doctests. 

I'm sure this was a mistake. I know some of the checks are supposed to not be required because they can fail a lot with unrelated failures (like the numpy nightly build that is currently failing), but I'm not completely sure which ones, which is why I haven't fixed this myself.

**Comment by oscarbenjamin** [comment]
It is not a mistake. I'm not sure there is much value in making everything required because often CI can fail for reasons unrelated to the PR in question. For example right now CI fails on all PRs due to gh-26862. There is no reason for that to prevent merging a PR. The fix is needed upstream (https://github.com/scipy/scipy/pull/21283 not sure what the hold up is).

The main things are required like that the test suite passes on current Python version, docs can build etc. It has been like this for some time and there haven't been any issues with a PR being merged that should have not been merged because some non-required job failed.

**Comment by asmeurer** [comment]
I agree that jobs like numpy nightly that can fail because of upstream issues should not be required. I'm talking about the tests and doctest builds. These are currently set as not required:

![image](https://github.com/user-attachments/assets/118ca9c5-3a06-4240-90ae-f22f52bd4b7a)
![image](https://github.com/user-attachments/assets/3faa1b48-1580-4eac-83e3-6c389c569326)

Not requiring these to pass before merging leads to situations where people merge broken PRs. Then you end up in a situation where every PR has failing CI, because like you said, the failure is "unrelated", so people just ignore the red ❌. And then more failures get merged because people don't notice them behind the existing ❌s. And you accumulate a bunch of test failures that then have to be fixed at release time. This the way things were 10+ years ago before GitHub implemented required checks. I can tell you that this is not a place where you want to be, especially as a release manager. 

Again, I would fix this myself, but I'm not completely sure which builds should be required or not. Certainly the base test and doctest should, but I don't know for instance if the PyPy, mpmath, flint, or benchmarks builds should. 

It would be nice if we could configure this directly in the yaml file instead of in the GitHub settings, so that we could notate why things are and aren't required. I guess the problem with that is anyone could just change it in a PR, although personally I think that's not really an issue.

**Comment by oscarbenjamin** [comment]
> I'm talking about the tests and doctest builds.

The tests and doctest jobs are required for the *current* Python version (i.e. 3.12, not shown in your screenshots). So are the slow tests and the docs build and a few others like authors and sdist that would otherwise cause problems at release time.

> Not requiring these to pass before merging leads to situations where people merge broken PRs.

As I said before it has been like this for a long time and this has not happened.

The actual work of keeping CI up and running is almost entirely done by me (e.g. gh-26862, https://github.com/scipy/scipy/issues/21272, https://github.com/numpy/numpy/issues/27054, https://github.com/scipy/scipy/pull/21283). The problems invariably come from external changes that are not prevented by restricting merging in the SymPy repo itself.

So far there are zero problems that result from a PR being merged because the checks were not all required. If that situation does arise then it will be easy to fix because we can just revert the PR.

The problems that arise at release time are all related to things that these CI checks do not cover anyway. I have added CI checks for the things that could easily be checked and that would otherwise be a problem at release time like the .mailmap file.

**Comment by asmeurer** [comment]
Why is only the latest Python version required? If we are going to test all Python versions, then we should also require them to pass. I can't think of an instance where an older Python version would fail and 3.12 would pass that wouldn't be something that should prevent a PR from being merged. Those builds aren't things that would fail because of upstream issues.

**Comment by oscarbenjamin** [comment]
> Why is only the latest Python version required?

The way that this is set up means that there is no need to edit the list of required jobs every time the Python versions change. Any job that has a version in the name is not required because it would mean needing to edit the list.

> If we are going to test all Python versions, then we should also require them to pass

We can require them to pass without marking them as required in the GitHub UI. If necessary we can enforce this by reverting a PR that was merged after some checks failed. Once again this has not happened though: the problems we are discussing here are hypothetical unlike the actual maintenance problems that always come from external changes.

This is not like the situation you refer to in the past when GitHub did not have required checks. It does have required checks and we do require some of the checks. The set of required checks is chosen so that the probability that a PR gets merged when it should not is very low (i.e. zero actual observed occurrences yet) but without me needing to edit the list of required checks continuously.

**Comment by asmeurer** [comment]
I can definitely sympathize with not wanting to deal with GitHub's UI for adding required checks. It's one of the worst UIs on GitHub.  

I think the reason it hasn't happened is 1) people are already sensitized to not merging PRs unless everything is green, because we've had required checks for so long (it would be nice if we could actually make the expected failures like numpy nightly less prominent, since they scare both contributors and unwary maintainers alike), and 2) a failure for only an older Python version is relatively rare. The only way it really happens is if someone uses a newer Python feature, and Python hasn't had many new features that are useful to SymPy or that people even know about.  

Even so, I think it is valuable to make these checks required, since they are actually required.

**Comment by oscarbenjamin** [comment]
> Even so, I think it is valuable to make these checks required, since they are actually required.

As I said before, they can be "required" without us marking them as "required" in the GitHub checks. It is just a case of us requiring them to pass which means potentially a revert if something breaks. It is already the case that we might need to revert after something breaks even if CI checks pass at the time. What is actually "required" by the project is never something that can be fully captured by ticking a box in GitHub to say that some CI jobs are "required". The notion of what is required needs to be understood in a broader sense than what can be fully captured by any CI script.

I think that the set of required checks is fine. We can reconsider this issue when it becomes something more than just hypothetical.

**Comment by sylee957** [comment]
Meanwhile, I also question the purpose of having sage test at least, because it seems like broken and never passing

`Run Sage CI / docker / linux (debian-bullseye, standard) (push) Failing after 72m`

**Comment by oscarbenjamin** [comment]
The problem with the Sage test is that the doctests needed to be updated because of changes in SymPy. I didn't want to update the Sage doctests after the change in gh-25673 so the Sage job has been broken since then. I assume that Sage developers will fix the doctests now after SymPy 1.13 was released but we might need to make the Sage CI point at a different Sage PR.

**Comment by asmeurer** [comment]
There's basically three types of CI jobs:

1. Basic tests. This includes the tests and doctests builds. These should always pass. It is very unlikely that these would break for a reason unrelated to a PR. 
2. Tests involving dependencies. These should also pass, but it's also quite likely that the dependency itself could break the test from underneath us. 
3. Tests of upstream packages. This includes the sage tests. The point here is more to make sure we aren't breaking an upstream package than the other way around. 

The first type should be required and a PR should never be merged if it is failing. The second type really should also always be passing, but because they can fail for unrelated reasons, they aren't required. The third type is just there for us to keep an eye on. If the upstream failure is "our fault" we need to fix it, but it could also be the upstream's fault. 

Back in the day Travis CI had "allowed failure" which let you make it so that jobs would not give the scary red :X: if they failed, but would still let you know that they were failing. GitHub Actions hasn't implemented this, even though many people have requested it https://github.com/actions/runner/issues/2347. So for now, you just have to know as a SymPy maintainer which :x: jobs you can ignore and which you can't. The "required" thing is the only indicator that exists to at least tell you which jobs are in the first category, but there's no easy way to indicate the second and third types (I guess the difference is the third type only runs on master, not PRs). Maybe we can work around this in other ways like putting "(allowed to fail)" in the job name. Another extreme option would be to make every job required and only un-requiring them when they start failing, but I think you and I are the only ones who can do that, so I understand not wanting to do that.

@oscarbenjamin, this is why having these jobs be actually "required" is important. It gives a signal to the SymPy maintainers who don't have the context that you have about which jobs are important to consider before merging and which aren't. The more :X: jobs which have to be ignored, the more important this becomes. People are starting to learn that you can ignore :x:'s in builds as long as the merge button is still there. 

It seems you are unwilling to do this, I guess because you don't want to deal with GitHub's interface. I'm happy to do it for you. As I mentioned in the OP, the main reason I haven't yet is I wanted to make sure I had clarification on which jobs are actually required or not. The fact that I don't know this should be reason enough to do this. If I don't know for certain which jobs are in which of the above categories, then other SymPy maintainers won't either.

**Comment by oscarbenjamin** [comment]
> I'm happy to do it for you.

Go ahead, but it needs updating continuously and I don't intend to do that the next time the jobs change (besides unrequiring anything that needs to be unrequired to unblock CI).

> The more :X: jobs which have to be ignored, the more important this becomes.

There is exactly one :X: job right now that can be ignored. If it was previously marked as required then we would have marked it as unrequired by now. I have done what I can to fix it but we just have to wait (https://github.com/scipy/scipy/pull/21283#issuecomment-2261646999). 

> If I don't know for certain which jobs are in which of the above categories, then other SymPy maintainers won't either.

I think it's clear what is required: the PR must not cause new failures in CI. As far as I am concerned all jobs need to pass. It is just that in some cases the cause of the failure is external and therefore not related to any PR here. I don't know how to better handle that besides attempting to get the problem fixed upstream or otherwise disabling the job or marking the test as xfail. I don't think messing around with the required labels helps with anything.

We could just disable the job. That particular job was added in preparation for NumPy 2.0. It is potentially useful that this bug would be fixed upstream in SciPy before it gets out in a release of NumPy.

**Comment by asmeurer** [comment]
GitHub has a new "branch ruleset" feature which might be easier to manage than the old branch protection rules. https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets

**Comment by oscarbenjamin** [comment]
I've just read that but it doesn't seem to provide anything that changes the situation as far as this issue is concerned. I don't see the purpose of this issue i.e. what problem it is supposed to address.

The idea is that we should have more rules enabled to prevent merging that could break CI but when was the last time that CI was broken as a result of merging anything?

In other words what real problem is this issue supposed to solve?

Above you said:

> Not requiring these to pass before merging leads to situations where people merge broken PRs. Then you end up in a situation where every PR has failing CI

I'm sure that was a problem in the past but it is not a problem now and I don't think that we are at risk of falling into that problem again. What triggered this issue was the fact that there was a job failing in CI but the cause of that was external so restrictions on merging *in this repo* would not have helped. The only additional thing we could do that would help in that situation would have been to disable the test (which I did do in the end and then I had to re-enable it afterwards).

As I said above:

> I think that the set of required checks is fine. We can reconsider this issue when it becomes something more than just hypothetical.

To make that more explicit: I don't think that there is anything to do here and so this issue should be closed.

@sylee957's point about the Sage CI job is valid but should probably be handled in a separate issue.
